### ISSUE TITLE
Request for new game addition: Renegade X

### ISSUE THREAD

**Issue by egtechgeek** [CubeCoders/AMPTemplates#367]
# Module Request

## Game Information:

Game Title: Rengade X
How to obtain: https://wiki.totemarts.games/wiki/Linux_Guide#winetricks
Suppoted OSs: Officially, Windows. Unofficially Linux w/ WINE
Supports RCON/Console input? Uses UDK Engine, which should support some modified form of RCON.

## I confirm:

- [X] that I have searched for an existing module request for this application.
- [X] that the server is an official server distributed by the games publishers/developers.
- [X] that the server is available to the general public.
- [X] that the server can be run headless (without a GUI).
- [X] I have already attempted to create a configuration myself using the configuration tool at https://config.getamp.sh/

**Comment by PhonicUK** [comment]
I'm not seeing any reference to any kind of separate, dedicated server application?

**Comment by egtechgeek** [comment]
Based on what i've read here: https://wiki.totemarts.games/wiki/Server#Starting_up_your_server
It looks like the dedicated server application is downloaded and extracted as a part of the client, but after extracting and installing, it would appear that it can be run standalone.

**Comment by fffreak9999** [comment]
I would also like to see it available in here. The game client acts as the server client when you use the server command line argument 

E.g: udk.exe server CNC-Walls?AdminPassword={}?GamePassword={} -port=7777 

I can help you understand the configuration files also (as one of the oldest server hosting for the game and as a former developer of the RX game)

**Comment by Greelan** [comment]
It's the need to run the client and then shut it down that creates the issue. Doesn't lend itself to a headless installation. Unless you know another way, such as with command line flags

**Comment by fffreak9999** [comment]
It has RCON that can trigger a client shutdown using the exit command, you could also force terminate the process

**Comment by egtechgeek** [comment]
I have researched and dabbled with this over the last year, since my first post. The first time chain of events for the template script  would likely look something like this:
-deploy new docker container
-install wine, .net, and other pre-reqs
-download game launcher
-start vnc server/viewer to run launcher and download game files
-after game files are downloaded, server launch can run via script, and controlled via RCON

**Comment by fffreak9999** [comment]
If you need any help answering questions, feel free to reach out to me, I’ll answer what I can.

As for the lutris scripts I believe most of them don’t work that well, but people did manage to get them working a long time ago.

**Comment by Greelan** [comment]
This really won't be viable in AMP. AMP is not designed as a docker frontend where you get a tailored prebuilt image specifically for a particular server. You could do that yourself, but it won't be for general use

**Comment by PhonicUK** [comment]
The requirement that you use VNC to get it going makes it an absolute non-starter. The game devs should be pressured into building a proper tool.

**Comment by egtechgeek** [comment]
> If you need any help answering questions, feel free to reach out to me, I’ll answer what I can.
> 
> As for the lutris scripts I believe most of them don’t work that well, but people did manage to get them working a long time ago.

fffreak9999, Are you on the discord? Or is there any we we could collaborate on this without tying up this issue thread until we've actually got something workable?

**Comment by fffreak9999** [comment]
> > If you need any help answering questions, feel free to reach out to me, I’ll answer what I can.
> > As for the lutris scripts I believe most of them don’t work that well, but people did manage to get them working a long time ago.
> 
> fffreak9999, Are you on the discord? Or is there any we we could collaborate on this without tying up this issue thread until we've actually got something workable?

I am in the game’s Discord server or you can DM me at fffreak9999 on discord.

**Comment by egtechgeek** [comment]
> The requirement that you use VNC to get it going makes it an absolute non-starter. The game devs should be pressured into building a proper tool. Closing until such time.

PhonicUK, any chance that we can get this reopened? 
We have some new developments: 
-VNC no longer required.
-Lutris no longer required.
-Game server download, installation, and configuration via command line, successfully tested and confirmed working.
-I have an AMP template in the works that i am testing, i have a tech-support topic open on the discord requesting clarification of missing uid key.

**Comment by Greelan** [comment]
Do you have instructions for the current server setup you can link to? You're welcome to struggle with getting a template to work but it would be quicker for me to do it xD

**Comment by Greelan** [comment]
This wiki still suggests it is not really suited for automation in AMP: https://wiki.totemarts.games/wiki/Server

**Comment by egtechgeek** [comment]
That wiki is substantially outdated. It was a lot of trial and error, but i have many mental notes that i need to put to paper (or keyboard). This is my first ever attempt at making a template, so for testing i make the amp template for windows only, but i am running it locally on a debian virtual machine. egtechgeek/renx_amp are the base files for getting it to cooperate with amp headless. Would you be willing to join a discord group with myself and fffreak9999? he's been incredibly helpful with this, from the actual game side

**Comment by Greelan** [comment]
Nice work figuring out how to run it headless! Not to be a dick about it, but it would take me 10x the time to work you through the ins and outs of the generic module rather than just doing it myself 😀

**Comment by egtechgeek** [comment]
I have no problems at all if you want to do the generic module work. I need to put my linux commands on to paper, so that you know what needs to go into the generic template... And i also need to write a few bash scripts for the linux side to automate the installation and configuration of wine, mono, and DXVK

**Comment by Greelan** [comment]
Don't bother writing scripts, just outline what is needed

**Comment by Greelan** [comment]
I see you are packaging the game client on your own GH. That's a bit of a no-no with AMP integration due to copyright and integrity issues. Is there a way to run the installer headless to install in the desired location?

These are the sorts of issues with this server that presents challenges for AMP integration.

**Comment by egtechgeek** [comment]
> I see you are packaging the game client on your own GH. That's a bit of a no-no with AMP integration due to copyright and integrity issues. Is there a way to run the installer headless to install in the desired location?
> 
> These are the sorts of issues with this server that presents challenges for AMP integration.

The official game client is not packaged, nor used. the command line launcher was developed by another github user. I clone everything to my repo for testing and dev just to keep things organized. The final product will be using git submodules.

**Comment by Greelan** [comment]
What's in prerelease.zip then?

**Comment by egtechgeek** [comment]
> What's in prerelease.zip then?

Same contents as renx_amp/prerelease. It's currently only for windows, but it's the compiled cli launcher, compiled xdelta for the launcher, and the scripts required in order to headlessly run the launcher, download the game client from the publishers servers, and then launch the actual game.

**Comment by egtechgeek** [comment]
> Don't bother writing scripts, just outline what is needed

Will need some scripts to automate install of prerequisites and keep things headless.

Requirements
Debian 12:
-Wine 9.0 
-winetricks
-mono-complete
-PietJankbal/powershell-wrapper-for-wine

Sources
-pminten/totemarts-patcher-cli
-jmacd/xdelta-gpl
-PietJankbal/powershell-wrapper-for-wine

**Comment by Greelan** [comment]
Right, so you are using a third party tool to get around the limitations of the official installation process? Can you link the repo that you are getting this from?

**Comment by egtechgeek** [comment]
> Right, so you are using a third party tool to get around the limitations of the official installation process? Can you link the repo that you are getting this from?

pminten/totemarts-patcher-cli

**Comment by Greelan** [comment]
OK, there are much cleaner ways to do this on both windows and Linux. 

BTW, you start the client in the update stages to install the UDK files. How do you stop it?

**Comment by egtechgeek** [comment]
> OK, there are much cleaner ways to do this on both windows and Linux.
> 
> BTW, you start the client in the update stages to install the UDK files. How do you stop it?

That would be a question best suited for fffreak9999. Historically, i've just terminated the process

**Comment by egtechgeek** [comment]
> OK, there are much cleaner ways to do this on both windows and Linux.
> 
> BTW, you start the client in the update stages to install the UDK files. How do you stop it?

So, it isn't a full game client. It is, as it's named, a patcher. It downloads the UDK files, and then in the future, will check and update them as needed. The "patcher" automatically terminates after it's done what it needs to do. Were you asking about the udk executable?

**Comment by Greelan** [comment]
Yeah, the last update stage. Not great for the process just to be kept running

**Comment by egtechgeek** [comment]
you confused me a bit. udk.exe does not run at all with the patcher

**Comment by Greelan** [comment]
Your last update stage runs `UDK.exe` via `run_renx.ps1`

**Comment by egtechgeek** [comment]
> Your last update stage runs `UDK.exe` via `run_renx.ps1`

That is not part of updating, nor the patcher. It's simply an easy way for me to append the required arguments to start the game client in server mode. udk.exe cannot run headless without supplemental -args

**Comment by Greelan** [comment]
Right, you're running the server the wrong way in the template, which is what confused me

I will put together a proper template this weekend xD

**Comment by egtechgeek** [comment]
I would be grateful if i could see the raw template you create after you have, so that i can learn the template system better for the future. The Amp configuration generator export data for what i've done so far is in the github repo, so you can see configuration files, and some configuration and settings. I would like to expand more and the configuration and settings, once i know better how.

**Comment by Greelan** [comment]
Of course, all my dev templates are in my public repo.

I have a working version (for Windows at least) in `Greelan/AMPTemplates:dev`. You can add that repo to your config repos in AMP and try it out.

I can get the server to install and start on Linux, but it crashes with shader compiler errors. Seems I haven't got quite the right things installed in the wine prefix. A lot of the info out there is inconsistent, and old.

**Comment by Greelan** [comment]
And now it is running on Linux ...

**Comment by egtechgeek** [comment]
Updating remote source Greelan/AMPTemplates:dev
Cloning into 'Greelan-AMPTemplates-dev'...
fatal: not in a git directory

What did i miss?

**Comment by egtechgeek** [comment]
LIM - Failure to make API call to RenegadeX01: Connection refused (127.0.0.1:8081)

Is this error related to the template manageability?

**Comment by fffreak9999** [comment]
Ok, so having attempted to run this in a Windows Environment, AMP based running doesn't work, although the files are generated correctly and in the correct structure. If you attempt to run the game server from the file structure manually things work, but just not through AMP.
[Log Files.zip](https://github.com/user-attachments/files/17531183/Log.Files.zip)
I've attached a ZIP folder with a bunch of the log files I have which have been generated. 
If you need me to do / grab anything else, just let me know, I am more than willing to help.

**Comment by Greelan** [comment]
What version of windows?

**Comment by egtechgeek** [comment]
Windows Server 2023

**Comment by egtechgeek** [comment]
> And now it is running on Linux ...

I attempted to add some configurable settings to amp for the game in renegadexconfig.json, but it seems to break the template. Do they need to be added in one of the other files as well?

**Comment by Greelan** [comment]
Bad JSON probably. Show me?

**Comment by egtechgeek** [comment]
> Bad JSON probably. Show me?

https://github.com/egtechgeek/AMPTemplates/blob/main/renegadexconfig_Test.json

**Comment by Greelan** [comment]
Yeah, fix your JSON

**Comment by egtechgeek** [comment]
> Yeah, fix your JSON

Ah! Found a typo. Question, what "InputType" could be used with amp that would offer a multi-selection, to set the MapRotation options?

**Comment by Greelan** [comment]
AMP can't handle the format of that setting.

BTW, none of your other settings will work unless you also include the ini section in ParamFieldName

**Comment by egtechgeek** [comment]
> AMP can't handle the format of that setting.
Hopefully in the future, it'll be possible. It's hardly a deal breaker.

> BTW, none of your other settings will work unless you also include the ini section in ParamFieldName
All of the entries i've added have ParamFieldName values that match the line items in related ini files listed in renegadexmetaconfig.json

**Comment by Greelan** [comment]
You still need the associated ini section 

Have a look at the settings I included

**Comment by egtechgeek** [comment]
> You still need the associated ini section
> 
> Have a look at the settings I included

thank you! Looking good now. I was also rather foolish and didn't validate my json before pushing

**Comment by Greelan** [comment]
Once you get the config manifest as you want it, flick it over and I can include it in the template before merging it. I will give you credit in the kvp xD

I will probably do some changes to align it to other template manifests, but happy for you to do most of the work!

**Comment by egtechgeek** [comment]
> Once you get the config manifest as you want it, flick it over and I can include it in the template before merging it. I will give you credit in the kvp xD
> 
> I will probably do some changes to align it to other template manifests, but happy for you to do most of the work!

Gotta give credit to fffreak9999 as well. His knowledge of the game's inner working was incredibly valuable. 

Any luck getting the windows issue debugged?

**Comment by Greelan** [comment]
> Windows Server 2023

Some versions of windows server seem to have issues with running servers as Network Service (which is what AMP uses). Not clear the reason why. For the most part it has tended to be server 2016 and sometimes 2019, with 2022 generally being OK

In one case a user installed a package on 2022 which resolved the issue for them, can't recall what it was

**Comment by Greelan** [comment]
> Gotta give credit to fffreak9999 as well. His knowledge of the game's inner working was incredibly valuable.
> 
> 

The key thing for getting it into AMP was the patcher to streamline the installation process. The patcher GitHub is pretty clear on how it is used. Other than that it is pretty much the same as other UDK servers like Subsistence and Chivalry: Medieval Warfare

**Comment by egtechgeek** [comment]
> Once you get the config manifest as you want it, flick it over and I can include it in the template before merging it. I will give you credit in the kvp xD
> 
> I will probably do some changes to align it to other template manifests, but happy for you to do most of the work!

Since the merging process seems like a bit of a hassle, i'm going to populate every possible setting before flicking it over, to minimize how soon we may have to push updates/modifications.

**Comment by Greelan** [comment]
Hopefully not every one - there are a bazillion in those files lmao

**Comment by Greelan** [comment]
Merging process not really a hassle - you create updates, test it, create a PR, it is reviewed and one finalised it is merged

How quick the process is depends on how good the PR is xD

**Comment by egtechgeek** [comment]
> Hopefully not every one - there are a bazillion in those files lmao

Not all, no. Just all the common ones

**Comment by egtechgeek** [comment]
I am changing the subcategories in renegadexconfig.json to "Basic" "Additional" and "Advanced", to better organize the various options. Are you on board with that?

**Comment by Greelan** [comment]
It's preferable to have consistency with other templates. The common ones are Server and Gameplay, and then if there a bunch that make sense to be grouped together you might have others like Network or Discord or even Miscellaneous 

Having other subcategories is usually aligned with servers that have their own admin setup - the idea is to make it consistent so it is easier to use

For general UE servers that is not a thing

**Comment by egtechgeek** [comment]
I've added a LOT of configurable settings, all of which is commonly used in this game's community... With them all on one page, it just looked super messy and disorganized. I guess it's easy enough at this point to publish one version, and i can run my own version if it bothers my ocd severely lol

**Comment by Greelan** [comment]
Yeah, sure, I am not saying don't use Subcategories, just more user friendly to have some consistency across servers

I generally also try to order things in importance and most relevance

Ultimately it's down to what makes the most sense

I bet my OCD beats yours lol

**Comment by egtechgeek** [comment]
> Yeah, sure, I am not saying don't use Subcategories, just more user friendly to have some consistency across servers
> 
> I generally also try to order things in importance and most relevance
> 
> Ultimately it's down to what makes the most sense
> 
> I bet my OCD beats yours lol

How's "Server" "Gameplay" and "Advanced"?

Or, how about this. Can a "Hidden" Setting be conditional? IE: A toggle under server settings which, if enabled, will display the Advanced settings within the Server category?

**Comment by Greelan** [comment]
Sounds good

Unfortunately, no to conditional settings. I think I first asked for that maybe 2 years ago lmao

**Comment by egtechgeek** [comment]
I just pushed to my repo... looks decent on my amp server. should be good for the merge. Since you said it's not a hassle, i'll tinker some more with the layout and additional configurable options in the future.

**Comment by Greelan** [comment]
No. I haven't found any way for UDK-based servers to output to stdout. And according to the wiki, the server uses its own form of RCON, which is not Source RCON (and likely not any other format that AMP knows)

**Comment by egtechgeek** [comment]
Could amp's console periodically read the output of udk.exe's log file? some way to see what's going on would be good. I'll dig into the rcon stuff with fffreak9999

**Comment by Greelan** [comment]
Not currently

**Comment by egtechgeek** [comment]
should we consider trying to add JAJames/Jupiter-Bot into the template? Apparently, this is the official/original dev-created management tool for renegade x

**Comment by Greelan** [comment]
Given it runs as a separate executable, not really

BTW, in your settings you added TimeLimit rather than the CNC Time Limit setting on the wiki - what's the diff?

**Comment by egtechgeek** [comment]
I believe that the wiki is just extremely outdated. CNCTimeLimit used to be in UDKGame.ini, but it is NOT. The TimeLimit parameter can now only be found in UDKRenegadeX.ini

**Comment by Greelan** [comment]
Right. But UDKRenegadeX.ini has the other time limit settings, hence my question

**Comment by Greelan** [comment]
And that is actually the wiki I am referring to, or more specifically the more detailed settings page that is linked there: https://wiki.totemarts.games/wiki/Server_Settings

I'm adding all those, that's why I am asking xD

**Comment by egtechgeek** [comment]
For different game modes. DMModeTimeLimit is for Deathmatch game mode. CnCModeTimeLimit is for command and conquer mode, which is also the default used by TimeLimit.

I would ignore them, i assume they will ultimately be deprecated, because as far as i know, there is no server flag to change game modes.

**Comment by Greelan** [comment]
Right. I just assumed it would be based on the map, eg CNC or TDM

**Comment by fffreak9999** [comment]
Timelimit is CNC mode, and uses to differentiate between two separate game types “Marathon” and “AOW”

AOW is a timed mode which ends once the time has counted down and other conditions such as overtime has ended, Marathon is an infinite time mode (timelimit = 0)

**Comment by fffreak9999** [comment]
DMTimeLimit is specific for deathmatch as it has code in place to handle the game differently between map styles

**Comment by fffreak9999** [comment]
> Right. I just assumed it would be based on the map, eg CNC or TDM

You are correct, maps which are configured for DM mode are configured as such in the map file itself, they have the TDM prefix naming scheme

**Comment by fffreak9999** [comment]
Also the wiki is indeed out of date, most of what was written there was from nearly half a decade ago or longer

**Comment by egtechgeek** [comment]
fffreak9999 just confirmed my suspicion: Game mode is set by command line or map (as you said). It isn't a configurable setting in an ini file. 

So the template startup script would have to be different for different game modes, unless the -args after ./udk.exe are configurable without too much scripting

**Comment by fffreak9999** [comment]
You can’t really play it properly in deathmatch on CNC maps, it just defaults to AOW mode, with the timelimit being forcibly set to DMTimeLimit

**Comment by egtechgeek** [comment]
I've never seen other modes really used, tbh

**Comment by fffreak9999** [comment]
Either way, if you want information on all the configs, that’s where I excel, having dealt with server hosting since before I was a developer.

**Comment by fffreak9999** [comment]
> I've never seen other modes really used, tbh

They aren’t that popular tbh,

CNC mode (timed and marathon)
DM/TDM
DEF (defend vs waves of AI)
These are the maps prefixes which are relevant since NVN was removed.

**Comment by Greelan** [comment]
> Either way, if you want information on all the configs, that’s where I excel, having dealt with server hosting since before I was a developer on the game. (10+ years ago I think)

Nice. How about this? Send me a list of the settings from each ini you think would be helpful in the UI, and I will do the template accordingly.

**Comment by Greelan** [comment]
> So the template startup script would have to be different for different game modes, unless the -args after ./udk.exe are configurable without too much scripting

Could potentially be done. Let me know what the options are

**Comment by egtechgeek** [comment]
I already have all of the create settings in my private template, but i put them in a separate config category as i said, because it made the game settings just look like a nightmare, in my opinion.

**Comment by egtechgeek** [comment]
> > So the template startup script would have to be different for different game modes, unless the -args after ./udk.exe are configurable without too much scripting
> 
> Could potentially be done. Let me know what the options are

Chatting with fffreak9999, better to leave as based on map selection. Not all maps support multiple modes, so until AMP allows for conditional fields, probably best to leave this out to prevent glitching.

**Comment by fffreak9999** [comment]
It was mentioned to me, I should explain about the mutator loading via the command line arguments

udk.exe server CNC-Under?mutator/s=blah.blah,blah2.blah2 is the default method, 
But there are a bunch which need the game mode which would be 
udk.exe server CNC-Under?Game=renx_game_pug.renx_game_pug

It depends on the invoking class from the mutator

**Comment by fffreak9999** [comment]
FYI: Was able to test with Windows 11 Pro as the OS.
The server was able to be hosted fine, no console output to the web page, but was joinable and configurable.

My Windows Server 2023 install didn't like running it from AMP, but the files themselves were usable outside of AMP.

**Comment by IceOfWraith** [comment]
Do you mean Server 2022? Some games don't work as NETWORK SERVICE, unfortunately. There's not much we can do about that. I believe 2022 is the one that actually allowed more to work though.

**Comment by fffreak9999** [comment]
Yeah 2022 server

**Comment by Greelan** [comment]
> Do you mean Server 2022? Some games don't work as NETWORK SERVICE, unfortunately. There's not much we can do about that. I believe 2022 is the one that actually allowed more to work though.

Wonder if running Process Monitor while the server is being started could reveal where it is getting unstuck? Could be an additional perm or package that could make the difference. Would be nice to solve this

**Comment by IceOfWraith** [comment]
I suspect it's more of a permission issue. AMP has similar issues accessing folders and things on Windows Server versus Pro. I've done some digging into it and got more clues that led me to that conclusion, but haven't figured out where the key is that unlocks the access it needs. Compared group policies, folder security, etc. No luck so far. But something is different in how NETWORK SERVICE is setup between the two for sure.

**Comment by egtechgeek** [comment]
I've been running successfully in AMP on Debian... Have you guys made any progress with Windows? Is this ready for merging?

**Comment by hypercube33** [comment]
> I've been running successfully in AMP on Debian... Have you guys made any progress with Windows? Is this ready for merging?

I dont see the config in the repo, maybe I'm blind but...I'd love to host RenX on my linux box as well. Thanks!
